Add NFT Design Doc (#1845)
hashgraph/hedera-mirror-node: docs/design/nft.md

#### Get Token by id

- Update `/api/v1/tokens/{id}` response to show NFTs by adding the `type` field.

```json
{
  "admin_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "decimals": "1000",
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "type": "NON_FUNGIBLE",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to return an error code 404 for `NON_FUNGIBLE` tokens that indicates
  users should go to `/api/v1/tokens/{id}/nfts`.

> 

Author: Updated this after discussing the fact that `/accounts` and `/balances` have issues now that we don't have the `nft_balance` table.  Effectively `tokens/{id}/nfts` is the balance endpoint for NFTs anyway.

  },
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "type": "NON_FUNGIBLE",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to return an error code 409 for `NON_FUNGIBLE` tokens that indicates
  users should go to `/api/v1/tokens/{id}/nfts`.

#### List NFTs

- GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.
  - NFTs should only display if the token has not been deleted (e.g. `token.deleted` is false) Otherwise, display empty
    list.
  - This endpoint should return a 409 for `FUNGIBLE` tokens with a message that indicates users should go
    to `/api/v1/tokens/{id}/balances`.

```json
{
  "nfts": [
    {
      "account_id": "0.0.111",
      "created_timestamp": "1610682445.003266000",

>       "deleted": true,

Author: That's fair, I hadn't considered how to handle that.  The two ways I can think are:
1. Add logic to the REST APIs to not show `account_id` when `deleted` is true (or show as null).
2. Have the importer mark `account_id` as null when deleted.

The first approach seems like the better one because the latter forces `account_id` in the DB to allow nulls, and we get back to the problem of `TokenDelete` having to do updates on every `nft` entry for it.  I've added some notes to the REST APIs for this, but I'm swapping the flag in this example to `false` just for simplicity.

>       "deleted": true,

Reviewer: flipping to `"deleted": false` should have addressed it.

The description above the example says:

"NFTs should only display if the token has not been deleted (e.g. token.deleted is false) Otherwise, display empty list.
account_id should not display when the NFT has been deleted."

Thus it's unnecessary to mark `account_id` as `null`

>       "deleted": true,

Author: Right, I meant that as an either-or situation.  I went with the first, just wanted to document my thoughts.

      "memo": "This is a test NFT",
      "modified_timestamp": "1610682445.003266001",
      "serial_number": 124
    }
  ],
  "links": {
    "next": null
  }
}
```

Optional Filters

> - `/api/v1/tokens/{id}/nfts?accountId=gt:0.0.111` - All NFTs belonging to a given account id.

Reviewer: ```suggestion
- `/api/v1/tokens/{id}/nfts?accountId=0.0.111` - All NFTs belonging to a given account id.
```

Not sure using `accountId` here is previously discussed or not, want to point out in other places it's `account.id`

Author: Removed the `gt` and fixed to `account.id`.

- `/api/v1/tokens/{id}/nfts?limit=x` - All NFTs taking the first `x` number of NFTs
- `/api/v1/tokens/{id}/nfts?order=desc` - All NFTs in descending order of `serial_number`
- `/api/v1/tokens/{id}/nfts?serialNumber=gt:0.0.1001` - All serial numbers in range

#### Get NFT by id

- GET `/api/v1/tokens/{id}/nfts/{serialNumber}` will show information about an individual NFT.

```json
{
  "account_id": "0.0.111",
  "created_timestamp": "1610682445.003266000",
  "deleted": false,
  "memo": "This is a test NFT",

>   "modified_timestamp": null,

Author: Good catch, removed the not null constraint.

>   "modified_timestamp": null,

Reviewer: umm, don't think we should drop `not null`

`modifed_timestamp` should have the same value as `created_timestamp` initially, then get updated whenever there's a change. this is aligned with `token.modified_timestamp`

>   "modified_timestamp": null,

Author: Oops, right, that makes sense.  Fixed, just added a timestamp to the REST example.

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Support hybrid NFTs (NFTs with a fungible child token to represent a fungible component)
- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

- Add to `token` table fields `type` (enum, values `FUNGIBLE` and `NON_FUNGIBLE`) and `max_supply` (bigint),
  - Default values will be `FUNGIBLE` and max long, respectively.

- Add a new `nft` table

```sql
create table if not exists nft
(
  account_id            bigint                  not null
  created_timestamp     bigint  primary key     not null,
  deleted               boolean default false   not null,
  modified_timestamp    bigint                  not null,
  memo                  text    default ''      not null,
  serial_number         bigint                  not null,
  token_id              bigint                  not null
);

```

- Add a unique constraint to `nft` for `token_id` and `serial_number`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp   bigint  not null,

>   nft_timestamp         bigint  not null,

Reviewer 2: Think you forgot to remove this field?

Author: Yep, removed now.

  },
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "decimals": "1000",
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "type": "NON_FUNGIBLE",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to return an error code 404 for `NON_FUNGIBLE` tokens that indicates
  users should go to `/api/v1/tokens/{id}/nfts`.

#### List NFTs

- GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.
  - NFTs should only display if the token has not been deleted (e.g. `token.deleted` is false) Otherwise, display empty
    list

>   - This endpoint should return a 404 if the token is not of type `NON-FUNGIBLE` with a message that indicates users

Reviewer 2: Perhaps a 409 is more appropriate

Author: That would line up a bit better, sure.  Changed to that here (and the token balance endpoint).

  - Have to determine if NFT or not
  - Now creates a Token and a list of NFTs
    - Should an NFT balance be created for treasury?

- Update Token
  - No change, unclear if NFT can be updated

Mint Token

- Create new NFT
  - Should an NFT balance be created for treasury?

Burn Token

- Have to delete NFT if present
- Unclear on how balance will be affected, if the serial number just disappears from the next balance file we may need
  to delete that entry as part of the burn

- Delete Token
  - NFTs are also deleted, this CANNOT be inherited because an NFT can be independently deleted

- Burn Token
  - Have to remove from balance, as file will not show

- Rename token_transfer to fungible_token_transfer
- Create non_fungible_token_transfer table
-
- Get tokens
  -

# Nonfungible Tokens (NFT)

## Purpose

NFTs allow users to create tokens that are indivisible, ensuring that each token is unique. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

> - Update `t_entity_types` to add`NFT`

Author: Questioning whether it is wise to make NFT an entity or keep everything in an `nft` table.  It's very dependent on the Token entity to get data, the only independent things would be the serial number/id itself, the memo, and the deleted flag (for burn/wipe cases).

Reviewer 3: I'm leaning to using the `TOKEN` entity for now and capturing fungibility maybe as a column

Author: That's fair, I have a `fungible` field already in the `token` table, I can add a `memo` and `deleted` field into the `nft` table and we won't have to deal with entities for NFTs at all.

Author: Made that change.

- Update `t_transaction_results` with new response codes
  - TBA

- Add to `token` table fields `maxSupply` (long) and `fungible` (boolean)
- Add a new `nft` table

```sql
create table if not exists nft
(
  serial_number         bigint              not null,
  nft_type_id           bigint              not null,
  hash                  bytea               not null,
  created_timestamp     bigint  primary key not null,
  modified_timestamp    bigint              not null,

>   memo

Author: Remove, part of entity

);

```

- Add a unique constraint to `nft` for `serial_number` and `created_timestamp`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  serial_number         bigint  not null,
  consensus_timestamp   bigint  not null,
  receiver_account_id   bigint  not null,

>   sender_account_id     bigint  not null,

Author: Debating if this is wise.  An NFT shouldn't show up twice in any transfer list, so we should be able to guarantee a sender and a receiver for each transfer.  It goes against the convention we have for crypto/token, but it would cut the rows in the table in half.

Reviewer 3: I'm actually not against this.
I think not having the other account on all our transfer tables makes things hard, hence the `non_fee_transfer` table.
This might be a change we should consider making eventually or starting here to help paint a more complete question when it comes to our transfer story.

);
```

- Add a unique constraint to `nft_transfer` for `serial_number` and `consensus_timestamp`, desc

- Add a new `nft_owner` table

```sql
create table if not exists nft_owner
(
  nft_id              bigint    not null,
  account_id          bigint    not null,
  consensus_timestamp bigint    not null,

> );

Author: Not set on a name. The convention says we stick with NftBalance, it's just a little odd since the balance of an NFT is either "has the one" or "has none"

Author: Also, need to check how we do the balance/account query.  If an NFT is burned/wiped, we may still pick it up since the account balance file likely will not overwrite this in any way.

```

- Add a unique constraint to `nft_transfer` for `serial_number` and `consensus_timestamp`, desc

### Importer

#### Converter

Add an `NftIdConverter`.

### Domain

- Add new fields to `Token` domain object from schema changes
- Add an `Nft` domain object with the same fields as the schema.
- Add an `NftTransfer` domain object with the same fields as the schema.

> - Add an `NftHolder` domain object with the same fields as the schema.

Author: Not set on this name.  The convention says we stick with `NftBalance`, it's just a little odd since the balance of an NFT is either "has the one" or "has none"

- Add an `NFT` enum value to `EntityTypeEnum`.
- Add `nftHolder` list to `AccountBalances` domain object

### Balance Parsing

Need information on file format. Effectively envision:

- Update `ProtoBalanceFileReader` to handle new NFT transfer list
  - Either deprecate use of the csv file or also add support for new CSV `BalanceFileReader` based on new format
  - Add `NftHolder` to the `AccountBalance` object as they are read.
- Update `AccountBalanceFileParser` to persist the `NftHolder`

### Entity Listener

- Add `onNft`
- Add `onNftTransfer`

### Entity Record Item Listener

- If transaction is successful, persist any NFT Transfers.
- `insertTokenCreate()` must be updated to create and persist the `NFT` objects and entities.
- `insertTokenMint()` must be updated to create and persist the `NFT` objects and entities.

> - `insertTokenBurn` must be updated to mark the NFT entities as deleted.

Author: Still not 100% clear on the inheritance logic for expiration, etc., but we can't inherit a deleted flag, because an NFT can be burned without the Token being deleted.

          }
        ]
      },
      "account": "0.0.8",
      "expiry_timestamp": null,
      "auto_renew_period": null,
      "key": null,
      "deleted": false
    }
  ],
  "links": {
    "next": null
  }
}
```

### Get Balances

- Update `/api/v1/balances` response to add token balances

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.0.8",
      "balance": 100,
      "tokens": []
    },
    {
      "account": "0.0.10",
      "balance": 100,
      "tokens": [
        {
          "token_id": "0.15.3",
          "balance": 80
        }
      ],
      "nftTypes": [
        {
          "id": "0.0.123",
          "nfts": [
            {
              "serialNumber": "0.0.124",
              "memo": "NFT1"
            },
            {
              "serialNumber": "0.0.124",
              "memo": "NFT2"
            }
          ]
        }
      ]
    },
    {
      "account": "0.0.13",
      "balance": 100,
      "tokens": [
        {
          "token_id": "0.15.3",
          "balance": 80
        },
        {
          "token_id": "0.2.4",
          "balance": 50
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

> ### List NFTTypes

Author: Remove.

# Non-Fungible Tokens (NFTs)

## Purpose

NFTs allow users to create tokens that each have unique data and can be individually tracked. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

- Add to `token` table fields `type` (enum, values `FUNGIBLE` and `NON_FUNGIBLE`) and `max_supply` (bigint),
  - Default values will be `FUNGIBLE` and max long, respectively.

- Add a new `nft` table

```sql
create table if not exists nft
(
  created_timestamp     bigint  primary key     not null,
  deleted               boolean default false   not null,
  modified_timestamp    bigint                  not null,

>   memo                  text    default ''      not null,+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Author: Yeah, just saw that, something was hitting that key on my keyboard.  Removed.

>   memo                  text    default ''      not null,+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Reviewer 2: Must've been a ferret

          "amount": -1200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.400",
          "amount": 1000,
          "token_id": "0.0.90000"
        }
      ],
      "transaction_hash": "aGFzaA==",
      "transaction_id": "0.0.10-1234567890-000000000",
      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "valid_duration_seconds": "11",
      "valid_start_timestamp": "1234567890.000000000"
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` and `/api/v1/accounts/{id}` response to add `nft_token_ids`
  - `nft_token_ids` should only show token_ids where the token was not deleted, and the account has at least one nft
    belonging to that token that has not been deleted (token.deleted | nft.deleted == false)

```json
{
  "accounts": [
    {
      "balance": {
        "account": "0.0.8",
        "auto_renew_period": null,
        "balance": 80,
        "deleted": false,
        "expiry_timestamp": null,
        "key": null,

>         "nft_token_ids": [

Reviewer 2: I don't think this is useful. Nor do I think it is possible as this endpoint is populated from balance tables but we removed the `nft_balance` table. Instead, we should utilize an account ID filter for `/api/v1/tokens` and `/api/v1/tokens/{id}/nfts`

Author: Agreed, didn't think about how tough this would be to construct without the `nft_balance` table.  The only thing I can think we're going to miss with this approach is there's not a great way to see historical "NFT balances" without hitting the `/nfts/transactions` endpoint for every NFT.  I don't think there's an actual ask for that, but it's the only thing we're missing for parity with existing HTS and accounts.

Author: Right now we have an account ID filter on `/tokens`, but it queries for tokens that are associated to the account.  For `/tokens/{id}/nfts?account.id` it will have to check the actual `nft` table to see which NFTs are currently owned by the account.  It might be slightly confusing that `/tokens` shows tokens, and then going to `/tokens/{id}/nfts` shows an empty list, but I don't know that we want to change the behavior of `/tokens`, would affect fungible tokens too.  I doubt associating TokenTypes and then not actually holding an NFT is going to be a common scenario anyway.

Author: Removed

          "0.2.9",
          "0.2.10"
        ],
        "timestamp": "0.000002345",
        "tokens": [
          {
            "balance": 80,
            "token_id": "0.15.3"
          },
          {
            "balance": 50,
            "token_id": "0.2.5"
          }
        ]
      }
    }
  ],
  "links": {
    "next": null
  }
}
```

#### Get Balances

- Update `/api/v1/balances` response to add `nft_token_ids` list
  - `nft_token_ids` should only show token_ids where the token was not deleted, and the account has at least one nft
    belonging to that token that has not been deleted (token.deleted | nft.deleted == false)

```json
{
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.0.8",
      "balance": 100,
      "tokens": []
    },
    {
      "account": "0.0.10",
      "balance": 100,

>       "nft_token_ids": [

Reviewer 2: I don't think this is useful. Nor do I think it is possible as this endpoint is populated from balance tables but we removed the `nft_balance` table. Instead, we should utilize an account ID filter for `/api/v1/tokens` and `/api/v1/tokens/{id}/nfts`

Author: Removed

Author: I also updated the `tokens/{id}/balances` section, as we'll have the same issue there.  Instead, we'll through an error there pointing to `/nfts`

      }
    }
  ]
}
```

Add optional filters

- `/api/v1/tokens?type=FUNGIBLE` - All fungible tokens (other values are `NON_FUNGIBLE` and `ALL` (default))

#### Get Token by id

- Update `/api/v1/tokens/{id}` response to show NFTs by adding the `type` field.

```json
{
  "admin_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "decimals": "1000",
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "type": "NON_FUNGIBLE",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

> - Update `/api/v1/tokens/{id}/balances` response to only show `account` when the token is an NftType token.\

Reviewer 2: `\` at end

Author: Fixed.

  - Having a list of serial_numbers could cause too much bloat

>   - Balances should only display if the token have not been deleted (`token.delted` == false)

Reviewer 2: ```suggestion
  - Balances should only display if the token has not been deleted (e.g. `token.deleted` is false).
```

Author: Fixed.

>     - An account balance should only be listed if the account has at least one nft that has not been
>       deleted. (`nft.deleted` == false)

Reviewer 2: ```suggestion
    - An account balance should only be listed if the account has at least one NFT that has not been
      deleted (e.g. `nft.deleted` is false).
```

Author: Fixed.

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.15.10"
    },
    {
      "account": "0.15.9"
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List NFTs

- GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.
  - NFTs should only display if the token have not been deleted (`token.delted` == false). Otherwise, display empty list
  - This endpoint should return a 404 if the token is not of type `NON-FUNGIBLE`

```json
{
  "nfts": [
    {
      "created_timestamp": "1610682445.003266000",
      "deleted": true,
      "memo": "This is a test NFT",
      "modified_timestamp": "1610682445.003266001",
      "serial_number": 124
    }
  ],
  "links": {
    "next": null
  }
}
```

Optional Filters

- `/api/v1/nfts?limit=x` - All NFTs taking the first `x` number of NFTs
- `/api/v1/nfts?order=desc` - All NFTs in descending order of `serial_number`

> - `/api/v1/nfts?serialNumber=gt:0.0.1001` - All serial numbers in range

Author: Fixed.  Also added `accountId` filter.

#### Get NFT by id

- GET `/api/v1/tokens/{id}/nfts/{serialNumber}` will show information about an individual NFT.

```json
{

>   "created_timestamp": "1610682445.003266000",

Reviewer 2: Even if it's not stored in the table, should we show who currently owns the nft?

Reviewer 3: Yes

Author: Added this, also added it to the table after discussing with Donald the expected transfer volume.

  "serial_number": 124
}
```

#### Get NFT transaction history

- GET `/api/v1/tokens/{id}/nfts/{serialNumber}/transactions` will return minimal transaction information for each
  transaction involving a given NFT.

```json
{
  "transactions": [
    {
      "consensus_timestamp": "1618591023.997420022",
      "id": "0.0.19789-1618805680-742097949",
      "type": "TOKENBURN"
    },
    {
      "consensus_timestamp": "1618591023.997420021",
      "id": "0.0.19789-1618805680-742097948",
      "receiver_account_id": "0.0.11",
      "sender_account_id": "0.0.10",
      "type": "CRYPTOTRANSFER"
    },
    {
      "consensus_timestamp": "1618591023.997420020",
      "id": "0.0.19789-1618805680-742097947",
      "treasury_account_id": "0.0.5",
      "type": "TOKENMINT"
    }
  ],
  "links": {
    "next": null
  }
}
```

### Monitor

- Make changes to the `ExpressionConverter`
  - Add a new `nft` expression
    - Creating an NFT requires the `tokenType` field to be set to `NON-FUNGIBLE`. `TokenCreate` for an NFT also cannot
      create serial numbers as part of the initial supply, so a `TokenMint` should follow to have serial numbers created
      for the transaction suppliers.

>     - Serial numbers are auto-incremented per token, so minting a hardcoded or configured (one universal value) number

Reviewer 2: I prefer this approach over the more complex `nft.1.serialNo.1` approach. If you're using the expression created serial numbers in a token transfer or associate you're not going to need too many of them. You'll only need a lot if you're performance testing token create or token mint, which don't require the serial number.

Author: Sure, I agree, I figured the added customization wasn't worth the complexity but wanted to write it down in case others thought differently.  I will go ahead and remove the mentions of it.

Author: Removed this `alternatively` section.

- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

- Add to `token` table fields `type` (enum, values `FUNGIBLE` and `NON_FUNGIBLE`) and `max_supply` (bigint),
  - Default values will be `FUNGIBLE` and max long, respectively.

- Add a new `nft` table

```sql
create table if not exists nft
(
  account_id            bigint                  not null
  created_timestamp     bigint  primary key     not null,
  deleted               boolean default false   not null,
  modified_timestamp    bigint                  not null,
  memo                  text    default ''      not null,
  serial_number         bigint                  not null,
  token_id              bigint                  not null
);

```

- Add a unique constraint to `nft` for `token_id` and `serial_number`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp   bigint  not null,
  nft_timestamp         bigint  not null,
  receiver_account_id   bigint  not null,
  sender_account_id     bigint  not null,
);
```

> - Add a unique constraint to `nft_transfer` for `consensus_timestamp` and `nft_timestamp`, desc

Reviewer 3: Unique id could be the other other mentioned `consensus_timestamp, token_id, receiver_account_id, sender_account_id`

Reviewer 2: We don't want an index that covers all fields. Large indexes was the reason I was proposing a nft_timestamp.

Author: Agreed, I've put it back to `consensus_timestamp`, `token_id`, and `serial_number`

  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "type": "NON_FUNGIBLE",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to only show `account` when the token is an NftType token.\
  - Having a list of serial_numbers could cause too much bloat
  - Balances should only display if the token have not been deleted (`token.delted` == false)
    - An account balance should only be listed if the account has at least one nft that has not been
      deleted. (`nft.deleted` == false)

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.15.10"
    },
    {
      "account": "0.15.9"
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List NFTs

- GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.

>   - NFTs should only display if the token have not been deleted (`token.delted` == false). Otherwise, display empty list

Reviewer 3: ```suggestion
  - NFTs should only display if the token have not been deleted (`token.deleted` == false). Otherwise, display empty list
```

Reviewer 2: Dupe

Author: Fixed.

the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_entity_types` to add`NFT`
- Update `t_transaction_results` with new response codes
  - TBA

- Add to `token` table fields `maxSupply` (long) and `fungible` (boolean)
- Add a new `nft` table

```sql
create table if not exists nft
(
  created_timestamp     bigint  primary key not null,
  hash                  bytea               not null,
  modified_timestamp    bigint              not null,
  nft_type_id           bigint              not null,
  serial_number         bigint              not null,
);

```

- Add a unique constraint to `nft` for `serial_number` and `created_timestamp`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp   bigint  not null,
  receiver_account_id   bigint  not null,
  sender_account_id     bigint  not null,
  serial_number         bigint  not null,

> );

Reviewer 3: Add the `nft_id`.
Let's reduce the need for additional joins to get that later and make it easier where we do need a join

Author: Done.  I renamed `nft_id` and `nft_type_id` everywhere to just `token_id` as well, I didn't realize I was mixing all those terms together, really it should just be `token_id` and `serial_number` everywhere now. Also added this to the `nft_balance`.

- Update `ProtoBalanceFileReader` to handle new NFT transfer list
  - Either deprecate use of the csv file or also add support for new CSV `BalanceFileReader` based on new format
  - Add `NftBalance` to the `AccountBalance` object as they are read.
- Update `AccountBalanceFileParser` to persist the `NftBalance`

#### Entity Listener

- Add `onNft`
- Add `onNftTransfer`

#### Entity Record Item Listener

- If transaction is successful, persist any NFT Transfers.
- `insertTokenCreate()` must be updated to create and persist the `NFT` objects and entities.
- `insertTokenMint()` must be updated to create and persist the `NFT` objects and entities.
- `insertTokenBurn` must be updated to mark the NFT entities as deleted.
- `insertTokenDelete` must be added to mark the NFT entities as deleted.
- `insertTokenWipe` must be updated to mark the NFT entities as deleted.

### REST API

#### Get Transaction

- Update `/api/v1/transactions` response to add nft transfers

```json
    {
  "transactions": [
    {
      "consensus_timestamp": "1234567890.000000001",
      "charged_tx_fee": 7,
      "max_fee": "33",
      "memo_base64": null,
      "name": "TOKENTRANSFER",
      "node": "0.0.3",

>       "nftTransfers": [

Reviewer 3: nit: Naming style inconsistent with other keys
```suggestion
      "nft_transfers": [
```

Author: Good catch, had a few of those.  They should all be fixed now.

        {
          "senderAccount": "0.0.122",
          "receiverAccount": "0.0.121",

>           "serialNumber": "0.0.124"

Reviewer 3: We should probably also capture the `nft_id` / `token_id`

Author: Added `token_id`

>         }

Author: After confirming NFTs and FTs should show together in one list, this is the only place where I have them separate, mostly because of the way I designed the table to hold fewer rows.  I'm thinking it's better to stick to the convention though and follow the way `token_transfers` is structured, regardless of the extra rows.

      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "valid_duration_seconds": "11",
      "valid_start_timestamp": "1234567890.000000000"
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` response to add NFTs to the tokens list, with serial numbers instead of balance.

```json
{
  "accounts": [
    {
      "balance": {
        "account": "0.0.8",
        "auto_renew_period": null,
        "balance": 80,
        "deleted": false
        "expiry_timestamp": null,
        "key": null,
        "timestamp": "0.000002345",
        "tokens": [
          {
            "balance": 80,
            "token_id": "0.15.3"
          },
          {
            "balance": 50,
            "token_id": "0.2.5"
          },
          {
            "token_id": "0.2.9",
            "serial_numbers": [
              "0.2.10",
              "0.2.11"
            ]
          }
        ]
      }
    }
  ],
  "links": {
    "next": null
  }
}
```

> #### Get Balances

Author: Need to alphabetize the rest of the APIs tomorrow.

    logic will be needed for performant NFT transfers.
    - Initial thought is to have the supplier swap the sender and receiver each time to transfer the NFT back and forth.
      This could be done only on the NFT transfer or on the entire transfer. This would likely limit performance as the
      swap has to be synchronized (or we just let the transactions fail when double-transfers occur).
      - This would also require different signatures on the transaction, which would require reworking how we handle the
        transaction signing.
    - Alernatively, the `ExpressionConverter` could just generate a large amount of serial numbers, and
      the `CryptoTransferTransactionSupplier` could just transfer one at a time via a counter until it runs out. This
      would be a much simpler approach, but it has obvious limitations.

### Acceptance Tests

Add acceptance tests that verify all transactions are handled appropriately. This includes tests for

- createToken with an NFT
- mintToken with an NFT
- burnToken with an NFT
- deleteToken with an NFT
- wipeTokenAccount with an NFT
- cryptoTransfer with an NFT

# Outstanding Questions and Concerns:

> 1. General protobuf questions.

Author: Clean this up to be less vague.

# Nonfungible Tokens (NFT)

## Purpose

NFTs allow users to create tokens that are indivisible, ensuring that each token is unique. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_entity_types` to add`NFT`
- Update `t_transaction_results` with new response codes
  - TBA

- Add to `token` table fields `maxSupply` (long) and `fungible` (boolean)
- Add a new `nft` table

```sql
create table if not exists nft
(
  serial_number         bigint              not null,
  nft_type_id           bigint              not null,
  hash                  bytea               not null,
  created_timestamp     bigint  primary key not null,
  modified_timestamp    bigint              not null,
);

```

- Add a unique constraint to `nft` for `serial_number` and `created_timestamp`, desc

- Add a new `nft_transfer` table

```sql

> create table if not exists nft_transfer

Author: It would make sense to rename `token_transfer` to `ft_transfer` (or something similar) to make this less blurry.

Author: Sure.  It does appear `token_transfer` is now a catchall in the HIP for both, a fungible token transfer is now an `AccountAmount`.

          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "token_transfers": [
        {
          "account": "0.0.200",
          "amount": 200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.300",
          "amount": -1200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.400",
          "amount": 1000,
          "token_id": "0.0.90000"
        }
      ],
      "nftTransfers": [
        {
          "senderAccount": "0.0.122",
          "receiverAccount": "0.0.121",
          "serialNumber": "0.0.124"
        }
      ]
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` response to add nfts

```json
{
  "accounts": [
    {
      "balance": {
        "timestamp": "0.000002345",
        "balance": 80,
        "tokens": [
          {
            "token_id": "0.15.3",
            "balance": 80
          },
          {
            "token_id": "0.2.5",
            "balance": 50
          }
        ],
        "nftTypes": [
          {
            "id": "0.0.123",
            "nfts": [
              {
                "serialNumber": "0.0.124",

>                 "memo": "NFT1"

Author: Maybe change to hash. Or just remove entirely, I was just looking for something else to identify with other than strictly the serial number.

              },
              {
                "serialNumber": "0.0.124",
                "memo": "NFT2"
              }
            ]
          }
        ]
      },
      "account": "0.0.8",
      "expiry_timestamp": null,
      "auto_renew_period": null,
      "key": null,
      "deleted": false
    }
  ],
  "links": {
    "next": null
  }
}
```

#### Get Balances

- Update `/api/v1/balances` response to add token balances

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.0.8",
      "balance": 100,
      "tokens": []
    },
    {
      "account": "0.0.10",
      "balance": 100,
      "tokens": [
        {
          "token_id": "0.15.3",
          "balance": 80
        }
      ],
      "nftTypes": [
        {
          "id": "0.0.123",
          "nfts": [
            {
              "serialNumber": "0.0.124",

>               "memo": "NFT1"

Author: Maybe change to hash.  Or just remove entirely, I was just looking for something else to identify with other than strictly the serial number.

>               "memo": "NFT1"

Author: Maybe change to hash. Or just remove entirely, I was just looking for something else to identify with other than strictly the serial number.

> # Nonfungible Tokens (NFT)

Author: Good catch.  This seems to be a common issue, even the HIP title is "HTS Non Fungible support".

> # Nonfungible Tokens (NFT)

Author: Fixed.

## Purpose

NFTs allow users to create tokens that each have unique data and can be individually tracked. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

>   - TBA

Author: Sure, removed.

> - Add to `token` table fields `maxSupply` (long) and `fungible` (boolean)

Author: Added bullet for default values.

- Add a new `nft` table

```sql
create table if not exists nft
(

>   created_timestamp     bigint  primary key not null,

Author: I thought about this, but I figure A. that could always be pulled from the latest `nft_balance` entry and B. that field could be updated frequently, whereas the rest of the table I don't expect to see a lot of updates.  I'm not against adding it, but that was my logic.

>   created_timestamp     bigint  primary key not null,

Reviewer 2: I believe a token associate must still occur. So guess that would be the canonical source of the linkage.

>   created_timestamp     bigint  primary key not null,

Author: Spoke to Donald about expected transfer rates for NFTs after Nana brought up adding the same field, we are expecting a relatively low volume of transfers overall, and this would solve the issue of showing the owner of a newly minted NFT (no transfers), so I have added this field.

  deleted               boolean             not null,

>   hash                  bytea               not null,

Author: Yes, was added before that was concrete.  It looks like `metadata` is now `memo` for an NFT, so I will keep `memo` and remove this.

>   hash                  bytea               not null,

Author: Removed.

  modified_timestamp    bigint              not null,

>   memo                  text    default ''  not null,

Author: Originally I think this was leftover from when I wasn't sure if NFT was to be an entity or not.  The HIP now appears to show `memo` will represent the metadata of the NFT.

>   memo                  text    default ''  not null,

Reviewer 2: Yes, they renamed `metadata` to `memo` in the doc so it's clear now the source.

  serial_number         bigint              not null,
  token_id              bigint              not null,
);

```

- Add a unique constraint to `nft` for `serial_number` and `created_timestamp`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp   bigint  not null,
  receiver_account_id   bigint  not null,
  sender_account_id     bigint  not null,
  serial_number         bigint  not null,
  token_id              bigint  not null,
);
```

- Add a unique constraint to `nft_transfer` for `serial_number` and `consensus_timestamp`, desc

- Add a new `nft_balance` table

```sql
create table if not exists nft_balance
(
  account_id          bigint    not null,
  consensus_timestamp bigint    not null,
  serial_number       bigint    not null,

>   token_id            bigint    not null,

Author: Should all be removed now.

);
```

> - Add a unique constraint to `nft_transfer` for `serial_number` and `consensus_timestamp`, desc

Author: Yep, copy/paste error, added the correct constaint.

### Importer

#### Converter

> Add an `NftIdConverter`.

Author: Yep, previously it was to be a new entity, looks like that has changed.  There's quite a difference between the HIP and the design docs I built this off of, will need to do a refresh.

> Add an `NftIdConverter`.

Author: Removed.

#### Domain

- Add new fields to `Token` domain object from schema changes
- Add an `Nft` domain object with the same fields as the schema.
- Add an `NftTransfer` domain object with the same fields as the schema.
- Add an `NftBalance` domain object with the same fields as the schema.
- Add an `NFT` enum value to `EntityTypeEnum`.
- Add `NftBalance` list to `AccountBalances` domain object

#### Balance Parsing

Need information on file format. Effectively envision:

- Update `ProtoBalanceFileReader` to handle new NFT transfer list

>   - Either deprecate use of the csv file or also add support for new CSV `BalanceFileReader` based on new format

Author: Makes sense.  I rewrote it to mention to update it if the csv is updated, with a link to the questions section (with that question added).

  - Add `NftBalance` to the `AccountBalance` object as they are read.
- Update `AccountBalanceFileParser` to persist the `NftBalance`

#### Entity Listener

- Add `onNft`
- Add `onNftTransfer`

#### Entity Record Item Listener

- If transaction is successful, persist any NFT Transfers.
- `insertTokenCreate()` must be updated to create and persist the `NFT` objects and entities.
- `insertTokenMint()` must be updated to create and persist the `NFT` objects and entities.
- `insertTokenBurn` must be updated to mark the NFT entities as deleted.
- `insertTokenDelete` must be added to mark the NFT entities as deleted.
- `insertTokenWipe` must be updated to mark the NFT entities as deleted.

### REST API

#### Get Transaction

- Update `/api/v1/transactions` response to add nft transfers

```json
    {
  "transactions": [
    {
      "consensus_timestamp": "1234567890.000000001",
      "charged_tx_fee": 7,
      "max_fee": "33",
      "memo_base64": null,
      "name": "TOKENTRANSFER",
      "node": "0.0.3",
      "nft_transfers": [
        {

>           "sender_account": "0.0.122",

Reviewer 2: Add `_id` to these

Author: Fixed.

          "receiver_account": "0.0.121",

>           "serial_number": "0.0.124",

Reviewer 2: number

          "token_id": "0.0.90000"
        }
      ],
      "transaction_hash": "aGFzaA==",
      "transaction_id": "0.0.10-1234567890-000000000",
      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "valid_duration_seconds": "11",
      "valid_start_timestamp": "1234567890.000000000"
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` response to add NFTs to the tokens list, with serial numbers instead of balance.

```json
{
  "accounts": [
    {
      "balance": {
        "account": "0.0.8",
        "auto_renew_period": null,
        "balance": 80,
        "deleted": false,
        "expiry_timestamp": null,
        "key": null,
        "timestamp": "0.000002345",
        "tokens": [
          {
            "balance": 80,
            "token_id": "0.15.3"
          },
          {
            "balance": 50,
            "token_id": "0.2.5"
          },
          {
            "token_id": "0.2.9",
            "serial_numbers": [
              "0.2.10",

>               "0.2.11"

Reviewer 2: Serial numbers are int64 not entity ids. So this should not be a string. Change everywhere else too.

Author: Fixed.

            ]
          }
        ]
      }
    }
  ],
  "links": {
    "next": null
  }
}
```

#### Get Balances

- Update `/api/v1/balances` response to add NFTs to the tokens list, with serial numbers instead of balance.

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.0.8",
      "balance": 100,
      "tokens": []
    },
    {
      "account": "0.0.10",
      "balance": 100,
      "tokens": [
        {
          "token_id": "0.15.3",
          "balance": 80
        },
        {

>           "token_id": "0.2.9",

Reviewer 2: We should add fungible flag to all to differentiate

Author: Added here, as well as `/transactions`.

      "admin_key": {
        "_type": "ED25519",
        "key": "31c4647554640c464c854337570217269a1fc0f8bc30591c349a410269090920"
      }
    },
    {
      "token_id": "0.0.10001",
      "symbol": "N",
      "fungible": false,
      "serial_numbers": [
        "0.0.1002",
        "0.0.1003"
      ],
      "admin_key": {
        "_type": "ED25519",
        "key": "31c4647554640c464c854337570217269a1fc0f8bc30591c349a410269090920"
      }
    }
  ]
}
```

#### Get Token by id

- Update `/api/v1/tokens/{id}` response to show NFTs by adding the `fungible` field and the `serial_numbers` list.

```json
    {
  "admin_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "decimals": "1000",
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "fungible": false,
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "serial_numbers": [
    "0.0.1002",
    "0.0.1003"

>   ],

Reviewer 2: An NFT is more than just its serial number. We should have a list of nfts documents that exposes data from nft table. Or if we think there will be too many to list in a sub document it should be a `/api/v1/tokens/{id}/nfts` API

Author: I debated this, I will look at it again after seeing what new fields we need based on the HIP.
I originally decided to make the separate `/nfts` endpoint instead because I could envision a scenario where someone only has the serial number and not the token_id, which would not be possible with this format, and at that point having all the nft data in this list as well seemed excessive, similar to how we don't pull full token data for `/tokens`, only for `/tokens/{id}`

Author: I think it would be bulky to add the memo/metadata to `/tokens`, but I've added the data to `/tokens/{id}` and all endpoints that branch from there.  Now that serial numbers aren't unique I've moved `/nfts` to `/tokens/{id}/nfts` as suggested, and added serialNumber as a possible filter on `/tokens`.

  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to show `serial_numbers` when the token is an NftType token.

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.15.10",
      "serial_numbers": [
        "0.0.1002",
        "0.0.1003"
      ]
    },
    {
      "account": "0.15.9",
      "serial_numbers": [
        "0.0.1004",
        "0.0.1005"
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List NFTs

- GET `/api/v1/nfts` will list basic information of all NFTs, including the NftType they belong to.

```json
{
  "nfts": [
    {
      "token_id": "0.0.123",
      "serial_number": "0.0.124",
      "memo": "NFT",
      "created_timestamp": "1610682445.003266000"
    }
  ],
  "links": {
    "next": null
  }
}
```

Optional Filters

> - `/api/v1/nfts?serialNumber.id=gt:0.0.1001` - All serial numbers in range

Reviewer 2: drop `.id`

Author: Removed.

- `/api/v1/nfts?token.id=0.0.1000` - All NFTs belonging to the given token type.
- `/api/v1/nfts?order=desc` - All NFTs in descending order of `serial_number`
- `/api/v1/nfts?limit=x` - All NFTs taking the first `x` number of NFTs

#### Get NFT by id

> - GET `/api/v1/nfts/{serialNumber}` will show more detailed information about an individual NFT.

Reviewer 2: Is a serial number globally unique or only unique within its token? If the latter does it start at 0 and increment by one or is it random?

Author: The original design documents had it as a globally unique id, but that was when it was to be its own entity.  I will enquire about how the serial number is now being handled, I had the same question when reading the HIP.

Author: Confirmed they are incremented and only unique per token, adjusted to account for that.

```json
{
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "hash": "123123123123A",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "memo": "MY FIRST NFT",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "tokenId": "0.0.123",
  "serialNumber": "0.0.124",
  "memo": "NFT",
  "created_timestamp": "1610682445.003266000",
  "wipe_key": {

>     https: //hedera.zoom.us/j/8826097631?pwd=VUxmeW13Y0luREJuWmwxRFpHZzQvUT09

Author: Fixed.

    logic will be needed for performant NFT transfers.
    - Initial thought is to have the supplier swap the sender and receiver each time to transfer the NFT back and forth.
      This could be done only on the NFT transfer or on the entire transfer. This would likely limit performance as the
      swap has to be synchronized (or we just let the transactions fail when double-transfers occur).
      - This would also require different signatures on the transaction, which would require reworking how we handle the
        transaction signing.
    - Alernatively, the `ExpressionConverter` could just generate a large amount of serial numbers, and
      the `CryptoTransferTransactionSupplier` could just transfer one at a time via a counter until it runs out. This
      would be a much simpler approach, but it has obvious limitations.

### Acceptance Tests

Add acceptance tests that verify all transactions are handled appropriately. This includes tests for

- createToken with an NFT
- mintToken with an NFT
- burnToken with an NFT
- deleteToken with an NFT
- wipeTokenAccount with an NFT

> - cryptoTransfer with an NFT

Reviewer 2: This is too generic. We should have an acceptance test for concrete flows that users will invoke. Flows like create, mint, transfer.

Author: Added details here, two new test suites to test the additional NFT logic, the rest should already be covered in HTS tests.

    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "type": "NON_FUNGIBLE",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to show `serial_numbers` when the token is an NftType token.

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.15.10",
      "nfts": [
        {
          "serial_number": 1002,
          "memo": "This is a test NFT"
        },
        {
          "serial_number": 1003,
          "memo": "This is another test NFT"
        }
      ]
    },
    {
      "account": "0.15.9",
      "nfts": [
        {
          "serial_number": 1004,
          "memo": "This is a third test NFT"
        },
        {
          "serial_number": 1005,
          "memo": "This is a fourth test NFT"
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List NFTs

> - GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.

Reviewer 3: You should call out the response if this is called on a fungible token. Probably 404 as an empty list could mean serials have not been minted or the ones that were got deleted and future ones could still be minted

Author: Good call, added notes what to do for `FUNGIBLE` tokens as well as deleted tokens

      need to mint a serial number for `nft.1` (after the creation of `nft.1`). This would require more work to refactor
      the `ExpressionConverter`, but it would allow for more customization.
  - Update the `token` expression logic to set the new fields (`tokenType` and `maxSupply`).

- Update the `TransactionSuppliers`
  - `TokenCreateTransactionSupplier` will need an enum `tokenType` attribute (with values FUNGIBLE and NON_FUNGIBLE) and
    a long `maxSupply` attribute.
    - Some fields will now have stricter requirements now, such as for NFTs `decimals` has to be 0, as
      does `initialSupply`. We could simply abide by the user config values, regardless of if they are valid, but most
      likely we should have logic to enforce those requirements based on the `tokenType`.
  - `TokenMintTransactionSupplier` will need an enum `tokenType` attribute, as well as a String `memo` attribute. The
    supplier should only set `amount` for fungible tokens, and only `memo` for NFTs
    - Alternatively, the `tokenType` can be removed, and the supplier just sets whichever of `memo` or `amount` is set,
      but this relies on users configuring things correctly.

>   - `TokenBurnTransactionSupplier` and `TokenWipeTransactionSupplier` will need a enum `tokenType` attribute, and it

Reviewer 3: nit:
```suggestion
  - `TokenBurnTransactionSupplier` and `TokenWipeTransactionSupplier` will need an enum `tokenType` attribute, and it
```

Author: Fixed.

    will need to set the `serialNumbers` list for NFTs (hardcoded based on the `initialSupply` value used in
    the `ExpressionConverter`)
    - The list could be configurable using the compound expression mentioned earlier if desired.

- Add support for transfering NFTs in `CryptoTransferTransactionSupplier`.

>   - Add a new `tokenType` enun attribute to be used when doing a TOKEN or BOTH transfer, that determines whether to use

Reviewer 3: ```suggestion
  - Add a new `tokenType` enum attribute to be used when doing a TOKEN or BOTH transfer, that determines whether to use
```

Author: Fixed.

    the `amount` attribute or the `serial numbers` list.
  - Because a serial number can only be transferred once out of a given account (unless it is transferred back), custom
    logic will be needed for performant NFT transfers.
    - One approach would be to have the supplier swap the sender and receiver each time to transfer the NFT back and
      forth. This could be done only on the NFT transfer or on the entire transfer. This would likely limit performance
      as the swap has to be synchronized (or we just let the transactions fail when double-transfers occur).
      - This would also require different signatures on the transaction, which would require reworking how we handle the
        transaction signing.
    - Alernatively, the `ExpressionConverter` could just generate a large amount of serial numbers, and
      the `CryptoTransferTransactionSupplier` could just transfer one at a time via a counter until it runs out. This
      would be a much simpler approach, but it would cap the number of transactions possible in a scenario, and it would
      require a longer startup time to generate all the serial nunmbers.

### Acceptance Tests

> Add acceptance tests that verify all transactions are handled appropriately. This includes tests for

Reviewer 3: You'll wanna add a test case for scheduled nft transfer

- Create a token with type `NON-FUNGIBLE`
 - Mint a serial number
 - Create an account with receiverSigRequired=true
 - Associate the token to an account
 - Create a schedule of CRYTOTRANSFER to transfer the nft
 - Submit a ScheduleSign from the receiving account
 - Verify response codes and data from `/tokens`, `/tokens/{id}/nfts`,
   `/tokens/{id}/nfts/{serialNumber}/transactions`, `/schedules/{id}`, and `/transactions`

Author: Thanks for writing it out for me, saved me a few minutes!  Added.

1. Basic NFT transfer flow

- Create a token with type `NON-FUNGIBLE`
- Mint a serial number
- Associate the token to an account
- Transfer the serial number to the account.
- Verify response codes and data from `/tokens`, `/tokens/{id}/nfts`,

>   `/tokens/{id}/nfts/{serialNumber}/transactions`, `/balances`, and `/transactions`,

Reviewer 3: Can't rely on balances because of the 15 min balance stream file frequency.
```suggestion
  `/tokens/{id}/nfts/{serialNumber}/transactions`, and `/transactions`,
```

Reviewer 3: Same below

Author: Removed in both.

      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "valid_duration_seconds": "11",
      "valid_start_timestamp": "1234567890.000000000"
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` and `/api/v1/accounts/{id}` response to add nfts list.

```json
{
  "accounts": [
    {
      "balance": {
        "account": "0.0.8",
        "auto_renew_period": null,
        "balance": 80,
        "deleted": false,
        "expiry_timestamp": null,
        "key": null,
        "nfts": [
          {
            "token_id": "0.2.9",
            "serial_numbers": [
              210,
              211
            ]
          }
        ],
        "timestamp": "0.000002345",
        "tokens": [
          {
            "balance": 80,
            "token_id": "0.15.3"
          },
          {
            "balance": 50,
            "token_id": "0.2.5"
          }
        ]
      }
    }
  ],
  "links": {
    "next": null
  }
}
```

#### Get Balances

> - Update `/api/v1/balances` and `/api/v1/balances/{id}` response to add nfts list

Reviewer 3: Do you mean to create a new `/api/v1/balances/{id}`, we should call that out explicitly ?

Author: No, just me mixing up the endpoints, I sometimes forget the details in `/accounts` vs `/balances`.  I've corrected this now.

# Non-Fungible Tokens (NFTs)

## Purpose

NFTs allow users to create tokens that each have unique data and can be individually tracked. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

- Add to `token` table fields `token_type` (enum, values `FUNGIBLE` and `NON_FUNGIBLE`) and `max_supply` (bigint),
  respectively.
  - Default values will be `true` and max long.

- Add a new `nft` table

```sql

> create table if not exists nft

Reviewer 3: Do we view this as a subset of the token table? similar to token_account?
In that case should we also have a column for `account_id` to capture who the nft belongs to?

Author: Yes, this would hold the data for the individual nfts that can't be represented in the `token` table (pretty much memo, the deleted flag, and some timestamps).  By `account_id` you mean the current holder?  Steven and I have been discussing that above, my original thought was that could be a frequently changing column in an otherwise mostly static table, and would better be left to `nft_balances`.  That table has since been removed per a suggestion we rely solely on the `nft_transfer` table to find that.

Reviewer 3: Yes `account_id` would be the current holder/owner. 
Right having that would mean every time a transfer happens we would have to update this table!

If we rely solely on `nft_transfer` then we'd always be searching the `nft_transfer` table to find the latest account that received the nft (by token_id & serial_number).

I guess it comes down to how individual nfts are traded. 
If it's rarely then having it here makes sense, if it's often then not having to update this table every time is a benefit. Really comes down to the application. In a collectors application trades are rare once acquired. In say a DEX application with high trading the swaps can get crazy.

Hmmm

Author: I've asked Donald for some clarity on the general expectations of NFT transfer rates.
The field would help solve the issue of initial ownership, we would set the `nft` account_id to the treasury account on create and would not have to do a search for an `nft` with no transfers.  Since you and Steven have both mentioned it, and it does help with a variety of issues, I think I'm okay with adding it, just want to see if I can get any specs that push the issue one way or the other.

Author: Donald confirmed we're expecting low volumes of transfers.  I will add an `account_id` field.

Author: Added

(
  created_timestamp     bigint  primary key     not null,
  deleted               boolean default false   not null,
  modified_timestamp    bigint                  not null,
  memo                  text    default ''      not null,
  serial_number         bigint                  not null,
  token_id              bigint                  not null
);

```

- Add a unique constraint to `nft` for `token_id` and `serial_number`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp   bigint  not null,

>   nft_timestamp         bigint  not null,

Reviewer 3: What timestamp is this? The nft `created_timestamp`?

Reviewer 3: Looking at info from this table would tell me stuff is being transferred, from and to whom. However, it doesn't tell me what. I would have expected the `token_id` and `serial_number`, is this an optimization on space?

Author: Yes, Steven suggested we remove `token_id` and `serial_number` because they are only unique in combination, it was making the index more complex, and replacing it with the timestamp of the nft's creation.  He also requested I shorten the name to that from what it was (`nft_created_consensus_timestamp`).  Willing to discuss this again, but that's how this came to be.

Reviewer 3: Hmmm, this now requires a forced join to figure out which nft details.
I think we can get away with this as is but once we iron out and expose a transfers story we'll realize this adds overhead.
Personally would prefer to have kept `token_id` and `serial_number`columns and make the index the complex (`consensus_timestamp, token_id, serial_number`) or (`consensus_timestamp, token_id, receiver_account_id, sender_account_id`) I prefer the former.

If we have to leave it as is I'd go with `nft_created_timestamp` over `nft_timestamp`, since there also exists a modified_timestamp for an nft

Author: I'll swap it to nft_created_timestamp for now, since I think all sides would agree to that at least.  You are correct that it forces a join on the `nft` table which is unfortunate.  It's a tradeoff of query complexity vs. index complexity. 
@steven-sheehy would be interested for your thoughts since you brought up the index concerns initially.

Reviewer 2: After checking again, perhaps it might be better to just use `consensus_timestamp, token_id, serial_number`. We at least won't have receiver and sender in the index so it won't be every field. And our rest api shows token and serial so we'll need to join if we don't have it.

Author: Sure, sounds good, I'll swap it back.  This was a bigger issue with the `nft_balance` table because that was every field, but now that table is gone.

Reviewer 3: Great, I think this is the right approach and the easier one to manage.

  receiver_account_id   bigint  not null,
  sender_account_id     bigint  not null,
);
```

- Add a unique constraint to `nft_transfer` for `consensus_timestamp` and `nft_timestamp`, desc

### Importer

#### Domain

- Add new fields to `Token` domain object from schema changes
- Add an `Nft` domain object with the same fields as the schema.
- Add an `NftTransfer` domain object with the same fields as the schema.

> - Add an `NftBalance` domain object with the same fields as the schema.

Reviewer 3: I don't see an `nft_balance` table. Is that missing?

Author: It was removed earlier, just forgot to get rid of these.  They've been removed now.

# Non-Fungible Tokens (NFTs)

## Purpose

NFTs allow users to create tokens that each have unique data and can be individually tracked. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

> - Ingest NFT transactions and persist to the database

Reviewer 3: we should maybe explicitly call out the type of NFT's we're looking to support first(whole, fractional, singleton) and in the no goals the ones we're not

Author: I'll take a look at documenting that, but there's still some debate about which types services will support or even can support, so that's a little fluid at the moment.  Here's the [document that breaks that down a little](https://www.notion.so/hedera/IWA-Compatibility-41648a9fb56b4bdeb28ceb8b5f088d89)

Reviewer 3: Yep, thought we could just mention Whole, Singleton and Fraction and say Hybrid wasn't supported. But you're right if it's in flux let's leave it as is and out

Author: Added some comments in the goals for this.

- Update `/api/v1/transactions` response to add nft transfers

```json
    {
  "transactions": [
    {
      "consensus_timestamp": "1234567890.000000001",
      "valid_start_timestamp": "1234567890.000000000",
      "charged_tx_fee": 7,
      "memo_base64": null,
      "result": "SUCCESS",
      "transaction_hash": "aGFzaA==",
      "name": "TOKENTRANSFER",
      "node": "0.0.3",
      "transaction_id": "0.0.10-1234567890-000000000",
      "valid_duration_seconds": "11",
      "max_fee": "33",
      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "token_transfers": [
        {
          "account": "0.0.200",
          "amount": 200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.300",
          "amount": -1200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.400",
          "amount": 1000,
          "token_id": "0.0.90000"
        }
      ],
      "nftTransfers": [
        {
          "senderAccount": "0.0.122",

>           "receiverAccount": "0.0.121",

Author: Based on the comment earlier about how this table should be structured.

  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "fungible": false,
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",
  "serial_numbers": [
    "0.0.1002",
    "0.0.1003"
  ],
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### List NFTs

GET `/api/v1/nfts`

```json
{
  "nfts": [
    {
      "token_id": "0.0.123",
      "serial_number": "0.0.124",
      "memo": "NFT",
      "created_timestamp": "1610682445.003266000"
    }
  ]
}
```

Optional Filters

- `/api/v1/nfts?serialNumber.id=gt:0.0.1001` - All serial numbers in range
- `/api/v1/nfts?token.id=0.0.1000` - All NFTs belonging to the given token type.
- `/api/v1/nfts?order=desc` - All NFTs in descending order of `serial_number`
- `/api/v1/nfts?limit=x` - All NFTs taking the first `x` number of NFTs

### Get NFT by id

> GET `/api/v1/nfts/{serialNumber}`

Author: Should this show the current holder of the token?  Since an NFT can only be held in one account at a time, it might make sense to represent that here.

// Maybe we show holder?

```json
{
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "memo": "MY FIRST NFT",
  "supply_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "tokenId": "0.0.123",
  "serialNumber": "0.0.124",
  "memo": "NFT",
  "created_timestamp": "1610682445.003266000",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "modified_timestamp": "1618510697.682451000"
}
```

### Get NFT transaction history

GET `/api/v1/nfts/{serialNumber}/transactions`

```json
{
  "history": [
    {
      "consensus_timestamp": "1618591023.997420020",
      "transaction_type": "TOKENBURN"

>     },

Author: If NFTs are able to be updated (other than deleting one) this query might get tricky.  We can fetch the transfer list from nft_transfers and use that timestamp to get any transaction data we want.  Same goes for create, because we have a `created_timestamp`.  The `modified_timestamp` should line up with the transaction where it was deleted, if that did occur.  But we don't currently have a table that would note the times an NFT got updated.

# Non-Fungible Tokens (NFTs)

## Purpose

NFTs allow users to create tokens that each have unique data and can be individually tracked. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

> - Add to `token` table fields `tokenType` (enum, values FUNGIBLE and NON_FUNGIBLE) and `maxSupply` (long)

Author: Went with `token_type` as we already have `token_id` in this table, and it matches the protobuf.  I'm okay with shortening to `type` though, since we've discussed making columns less verbose.
The other suggestions here are fixed.

> - Add to `token` table fields `tokenType` (enum, values FUNGIBLE and NON_FUNGIBLE) and `maxSupply` (long)

Author: Changed it to `type`

>   - Default values will be `true` and max long.

Reviewer 2: ```suggestion
  - Default values will be `FUNGIBLE` and max long, respectively.
```

Author: Good call, added.

- Add a new `nft` table

```sql
create table if not exists nft
(
  created_timestamp     bigint  primary key     not null,
  deleted               boolean default false   not null,
  modified_timestamp    bigint                  not null,
  memo                  text    default ''      not null,
  serial_number         bigint                  not null,
  token_id              bigint                  not null
);

```

- Add a unique constraint to `nft` for `token_id` and `serial_number`, desc

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp               bigint  not null,

>   nft_created_consensus_timestamp   bigint  not null

Reviewer 2: I think this can be shortened to just `nft_id` or `nft_timestamp`.

Author: I like `nft_timestamp`, changed.

  receiver_account_id               bigint  not null,
  sender_account_id                 bigint  not null,
);
```

- Add a unique constraint to `nft_transfer` for `consensus_timestamp` and `nft_created_consensus_timestamp`, desc

- Add a new `nft_balance` table

```sql
create table if not exists nft_balance
(
  account_id                        bigint  not null,
  consensus_timestamp               bigint  not null,

>   nft_created_consensus_timestamp   bigint  not null

Reviewer 2: Same `nft_id` or `nft_timestamp`.

Author: Removed the table as per earlier discussion.

- Add an `NftBalance` domain object with the same fields as the schema.
- Add `NftBalance` list to `AccountBalances` domain object

#### Balance Parsing

Need information on file format. Effectively envision:

- Update `ProtoBalanceFileReader` to handle new NFT balance list
  - If the NFT balance list is added to the CSV version (see [questions](#outstanding-questions-and-concerns)), add
    support for new CSV `BalanceFileReader` based on new format
  - Add `NftBalance` to the `AccountBalance` object as they are read.
- Update `AccountBalanceFileParser` to persist the `NftBalance`

#### Entity Listener

- Add `onNft`
- Add `onNftTransfer`

#### Entity Record Item Listener

- If transaction is successful, persist any NFT Transfers.
- `insertTokenCreate()` must be updated to set `tokenType` and `maxSupply`
- `insertTokenMint()` must be updated to handle the new field `amountOrMemo` create and persist the `NFTs` if the memo
  is set.
- `insertTokenBurn` must be updated to handle the new field `amountOrSerialNumbers`, mark the `NFTs` as deleted, and
  update `modifiedTimestamp` if the `serialNumbers` list is set.

> - `insertTokenDelete` must be added to mark the `NFTs` as deleted and update `modifiedTimestamp` if the `tokenType`

Reviewer 3: I'm thinking we might want to give some thought to offloading some logic to the db layer for scenarios like this, that way we don't hold up the parser. e.g. triggers or some repeatable migrations that can be scheduled, I'm sure other efficient options exist.

> - `insertTokenDelete` must be added to mark the `NFTs` as deleted and update `modifiedTimestamp` if the `tokenType`

Author: You're correct that that was my plan, I was a bit worried about how that would affect the performance.  I like your idea, I'll remove this line and add a note in the REST API sections to do this logic.

  field is NON_FUNGIBLE for the token.

> - `insertTokenWipe` must be must be updated to handle the new field `amountOrSerialNumbers`, mark the `NFTs` as deletedm

Author: Fixed.

      "node": "0.0.3",
      "result": "SUCCESS",
      "token_transfers": [
        {
          "account": "0.0.200",
          "amount": 200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.300",
          "amount": -1200,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.400",
          "amount": 1000,
          "token_id": "0.0.90000"
        }
      ],
      "transaction_hash": "aGFzaA==",
      "transaction_id": "0.0.10-1234567890-000000000",
      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "valid_duration_seconds": "11",
      "valid_start_timestamp": "1234567890.000000000"
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` and `/api/v1/accounts/{id}` response to add nfts list.

```json
{
  "accounts": [
    {
      "balance": {
        "account": "0.0.8",
        "auto_renew_period": null,
        "balance": 80,
        "deleted": false,
        "expiry_timestamp": null,
        "key": null,

>         "nfts": [

Reviewer 3: Yeah, as we grow having it at the base level will become problematic cause the response size will bloat.
Might be a breaking change but worth it to move balances to the single account level.

If we limit the amount we also have to provide logic to page the entries to get the total list eventually or notify that this is a partial list and they have to go to `/api/v1/tokens/{tokenId}/balances?account.id=<accountId>` to get the full list.
Either ways requires some more planning

Notably we don't have a single balance resource endpoint

>         "nfts": [

Author: Been pondering this for a while now.  All of `/balances`, `/accounts`, and `/token/{id}/balances` could have this problem.  `/api/v1/accounts/{id}/nfts` could also have the same issue, if one token in the account has 10000 serial numbers and another has 10000 then splitting that up gets tricky.  It's really the combination of account_id and token_id that helps.

Nana's idea for `/api/v1/tokens/{tokenId}/balances?account.id=<accountId>` would work if we're comfortable with only showing serial information when a filter is applied.  Or we could do any of `/tokens/{tokenId}/balances/{accountId}`, or either of the two going from `/accounts` instead.  `/balances` would get messy, since we'd need both filters to do this, unless we do as Nana suggested and move it entirely under `/accounts`, which is a much bigger change.

I think still having the `nft` list with the token_ids would be good in these other places, but I will remove the serial_numbers list.

>         "nfts": [

Author: I've removed the serial number list from the 3 endpoints for the time being and left a note about that, still debating how/where to show the list.

>         "nfts": [

Author: From all the discussion downstream, the direction seems to be having users go to `/tokens` with an account id and the fungible flag `false`, and then going to each `/tokens/{id}/nfts` to get the detailed breakdown.  This should cover us for everything except historical balances, which I don't think is a major priority.

  ],
  "links": {
    "next": null
  }
}
```

#### Get Balances

- Update `/api/v1/balances` and `/api/v1/balances/{id}` response to add nfts list

```json
{
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.0.8",
      "balance": 100,
      "tokens": []
    },
    {
      "account": "0.0.10",
      "balance": 100,
      "nfts": [
        {
          "serial_numbers": [
            210,
            211
          ],
          "token_id": "0.2.9"
        }
      ],
      "tokens": [
        {
          "balance": 80,
          "token_id": "0.15.3"
        }
      ]
    },
    {
      "account": "0.0.13",
      "balance": 100,
      "nfts": [
        {
          "serial_numbers": [
            157
          ],
          "token_id": "0.15.6"
        }
      ],
      "tokens": [
        {
          "balance": 80,
          "token_id": "0.15.3"
        },
        {
          "balance": 50,
          "token_id": "0.2.4"
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List Tokens

- Update `/api/v1/tokens` response to show NFTs by adding the `tokenType` field.

```json
{
  "tokens": [
    {
      "token_id": "0.0.1000",
      "symbol": "F",

>       "tokenType": FUNGIBLE,

Reviewer 2: JSON errors here and elsewhere for this enum. Should have double quotes.

Author: Whoops, apparently I did not push the fix for that the other day.  That should be fixed everywhere now, also changed it to `token_type`.

Author: `token_type` now just `type` since it's in `token` context

      "admin_key": {
        "_type": "ED25519",
        "key": "31c4647554640c464c854337570217269a1fc0f8bc30591c349a410269090920"
      }
    },
    {
      "token_id": "0.0.10001",
      "symbol": "N",
      "tokenType": NON_FUNGIBLE,
      "admin_key": {
        "_type": "ED25519",
        "key": "31c4647554640c464c854337570217269a1fc0f8bc30591c349a410269090920"
      }
    }
  ]
}
```

Add optional filters

> - `/api/v1/tokens?type=fungible` - All fungible tokens (other values are `non_fungible` and `both` (default))

Author: Fixed and fixed.

> - `/api/v1/tokens?serial.number=1001` - All tokens that contain a serial number `1001` (implied that `type` will

Reviewer 2: ```suggestion
- `/api/v1/tokens?serialNumber=1001` - All tokens that contain a serial number `1001` (implied that `type` will
```

Author: Fixed.

> - `/api/v1/tokens?serial.number=1001` - All tokens that contain a serial number `1001` (implied that `type` will

Author: Sure.  I was envisioning a scenario where someone has the serial number but not the token id and needed the token data.  But that was not a requirement.  Removed.

  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to show `serial_numbers` when the token is an NftType token.

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.15.10",
      "nfts": [
        {
          "serial_number": 1002,
          "memo": "This is a test NFT"
        },
        {
          "serial_number": 1003,
          "memo": "This is another test NFT"
        }
      ]
    },
    {
      "account": "0.15.9",
      "nfts": [
        {
          "serial_number": 1004,
          "memo": "This is a third test NFT"
        },
        {
          "serial_number": 1005,
          "memo": "This is a fourth test NFT"
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List NFTs

- GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.

```json
{
  "nfts": [
    {
      "created_timestamp": "1610682445.003266000",
      "deleted": true,
      "memo": "This is a test NFT",
      "modified_timestamp": "1610682445.003266001",
      "serial_number": 124
    }
  ],
  "links": {
    "next": null
  }
}
```

Optional Filters

> - `/api/v1/nfts?deleted=false` - All serial numbers that have not been deleted.

Reviewer 2: Querying on just a boolean would be slow, hence why we don't support it for others. Let's remove for now.

Author: Makes sense, removed.

# Non-Fungible Tokens (NFTs)

## Purpose

NFTs allow users to create tokens that each have unique data and can be individually tracked. This document explains how
the mirror node can be updated to add support for NFTs.

## Goals

- Ingest NFT transactions and persist to the database
- Expose NFT info and transfers via REST APIs
- Monitor NFTs

## Non Goals

- Provide an NFT gRPC API

## Architecture

### Database

- Update `t_transaction_results` with new response codes

> - Add to `token` table fields `fungible` (boolean) and `maxSupply` (long)

Reviewer 2: Should we have an enum instead of a boolean to match the proto? Do we know if there ever could be a 3rd type?

Author: Sure, it could be possible considering they've talked about NFTs inside fungible tokens and vice-versa.  I've changed it.

>   - Default values will be `true` and `null` (null `maxSupply` will represent an infinite amount of tokens are possible)

Reviewer 2: It sounds like the default value for `maxSupply` should be max long per the doc's matrix and this comment:

> By default, if maxSupply is not provided, the token will be defined as having Infinite supply. (backwards compatible)

Author: Got it, yeah, the infinite concept threw me.  Changed it.

- Add a new `nft` table

```sql
create table if not exists nft
(
  created_timestamp     bigint  primary key not null,

>   deleted               boolean             not null,

Author: Fixed.

  modified_timestamp    bigint              not null,
  memo                  text    default ''  not null,
  serial_number         bigint              not null,
  token_id              bigint              not null
);

```

> - Add a unique constraint to `nft` for `token_id`, `serial_number`, and `created_timestamp`, desc

Author: Makes sense, fixed.

- Add a new `nft_transfer` table

```sql
create table if not exists nft_transfer
(
  consensus_timestamp   bigint  not null,
  receiver_account_id   bigint  not null,
  sender_account_id     bigint  not null,
  serial_number         bigint  not null,
  token_id              bigint  not null
);
```

> - Add a unique constraint to `nft_transfer` for `token_id`, `serial_number` and `consensus_timestamp`, desc

Author: Yeah, it's a bit hairy.  That's a clever idea, I've swapped those two fields out for `nft_created_consensus_timestamp`.  We could also do that for nft_transfer if want to further save space.

- Add a new `nft_balance` table

```sql

> create table if not exists nft_balance

Reviewer 2: We should also consider pushing back against NFTs being added to balances file. There's no actual balance being changed, it's just a list of serial numbers associated with a token. That info is already fully contained in the other tables unless someone has a start date before the token create. It's just duplication of information for us. If someone creates 10M NFTs for one token, it doesn't make sense to send 10M serial numbers every 15 min.

Reviewer 2: I would suggest we remove it, once it's confirmed it won't show up in the balance proto.

We don't need a `nft_owner` table as the `nft_transfer` provides the exact same information. It provides both the current and historical ownership of NFTs.

Author: Only concern would be an NFT with no transfers, i.e. an NFT that was just minted would be in the treasury balance, so would have to check for that, and check the `deleted` flag(s) and timestamps.  But that's doable, so yes, I will remove this.

Author: Removed this table

(
  account_id          bigint    not null,
  consensus_timestamp bigint    not null,
  serial_number       bigint    not null,
  token_id            bigint    not null
);
```

> - Add a unique constraint to `nft_balance` for `token_id`, `serial_number`, `account_id`, and `consensus_timestamp`,

Reviewer 2: Again, we can maybe add nft timestamp. We can also consider keeping serial_number and token_id in addition.

  desc

### Importer

#### Domain

- Add new fields to `Token` domain object from schema changes
- Add an `Nft` domain object with the same fields as the schema.
- Add an `NftTransfer` domain object with the same fields as the schema.
- Add an `NftBalance` domain object with the same fields as the schema.
- Add `NftBalance` list to `AccountBalances` domain object

#### Balance Parsing

Need information on file format. Effectively envision:

- Update `ProtoBalanceFileReader` to handle new NFT balance list
  - If the NFT balance list is added to the CSV version (see [questions](#outstanding-questions-and-concerns)), add
    support for new CSV `BalanceFileReader` based on new format
  - Add `NftBalance` to the `AccountBalance` object as they are read.
- Update `AccountBalanceFileParser` to persist the `NftBalance`

#### Entity Listener

- Add `onNft`
- Add `onNftTransfer`

#### Entity Record Item Listener

- If transaction is successful, persist any NFT Transfers.
- `insertTokenCreate()` must be updated to set `fungible` and `maxSupply`
- `insertTokenMint()` must be updated to handle the new field `amountOrMemo` create and persist the `NFTs` if the memo
  is set.

> - `insertTokenBurn` must be updated to handle the new field `amountOrSerialNumbers` and mark the `NFTs` as deleted if

Reviewer 2: And update modified timestamp

Author: Good catch, added here and for delete and wipe.

  the `serialNumbers` list is set.
- `insertTokenDelete` must be added to mark the `NFTs` as deleted if the `fungible` field is false for the token.
- `insertTokenWipe` must be must be updated to handle the new field `amountOrSerialNumbers` and mark the `NFTs` as
  deleted if the `serialNumbers` list is set.

### REST API

#### Get Transaction

- Update `/api/v1/transactions` response to add nft transfers

```json

>     {

Reviewer: nit: indentation

Author: Fixed.

  "transactions": [
    {
      "consensus_timestamp": "1234567890.000000001",
      "charged_tx_fee": 7,
      "max_fee": "33",
      "memo_base64": null,

>       "name": "TOKENTRANSFER",

Reviewer: ```suggestion
      "name": "CRYPTOTRANSFER",
```

Author: Fixed.

      "node": "0.0.3",
      "result": "SUCCESS",
      "token_transfers": [
        {
          "account": "0.0.200",
          "amount": 200,
          "fungible": true,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.300",
          "amount": -1200,
          "fungible": true,
          "token_id": "0.0.90000"
        },
        {
          "account": "0.0.400",
          "amount": 1000,
          "fungible": true,
          "token_id": "0.0.90000"
        },
        {

>           "fungible": true,

Reviewer: ```suggestion
          "fungible": false,
```

Author: Good catch, fixed

        }
      ],
      "transaction_hash": "aGFzaA==",
      "transaction_id": "0.0.10-1234567890-000000000",
      "transfers": [
        {
          "account": "0.0.9",
          "amount": 10
        },
        {
          "account": "0.0.10",
          "amount": -11
        },
        {
          "account": "0.0.98",
          "amount": 1
        }
      ],
      "valid_duration_seconds": "11",
      "valid_start_timestamp": "1234567890.000000000"
    }
  ]
}
```

#### Get Accounts

- Update `/api/v1/accounts` response to add NFTs to the tokens list, with serial numbers instead of balance.

```json
{
  "accounts": [
    {
      "balance": {
        "account": "0.0.8",
        "auto_renew_period": null,
        "balance": 80,
        "deleted": false,
        "expiry_timestamp": null,
        "key": null,
        "timestamp": "0.000002345",
        "tokens": [
          {
            "balance": 80,
            "fungible": true,
            "token_id": "0.15.3"
          },
          {
            "balance": 50,
            "fungible": true,
            "token_id": "0.2.5"
          },
          {
            "token_id": "0.2.9",
            "fungible": false,

>             "serial_numbers": [

Reviewer 2: nft balances also quite differ from fungible and we should consider a separate array. The current structure would be very hard for clients with strongly typed domain mapping tools to parse different fields depending upon a flag.

              210,
              211
            ]
          }
        ]
      }
    }
  ],
  "links": {
    "next": null
  }
}
```

#### Get Balances

- Update `/api/v1/balances` response to add NFTs to the tokens list, with serial numbers instead of balance.

```json

>     {

Reviewer: nit: indentation

Author: Fixed.

          "fungible": true,
          "token_id": "0.15.3"
        },
        {
          "fungible": false,
          "serial_numbers": [
            210,
            211
          ],
          "token_id": "0.2.9"
        }
      ]
    },
    {
      "account": "0.0.13",
      "balance": 100,
      "tokens": [
        {
          "balance": 80,
          "fungible": true,
          "token_id": "0.15.3"
        },
        {
          "balance": 50,
          "fungible": true,
          "token_id": "0.2.4"
        },
        {
          "fungible": false,
          "serial_numbers": [
            157
          ],
          "token_id": "0.15.6"
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List Tokens

- Update `/api/v1/tokens` response to show NFTs by adding the `fungible` field and the `serial_numbers` list.

```json
{
  "tokens": [
    {
      "token_id": "0.0.1000",
      "symbol": "F",
      "fungible": true,
      "admin_key": {
        "_type": "ED25519",
        "key": "31c4647554640c464c854337570217269a1fc0f8bc30591c349a410269090920"
      }
    },
    {
      "token_id": "0.0.10001",
      "symbol": "N",
      "fungible": false,

>       "serial_numbers": [

Reviewer 2: Remove

Author: Removed.  I assume I should also remove the `serialNumber` filter for this API?

Reviewer 2: Correct

        1002,
        1003
      ],
      "admin_key": {
        "_type": "ED25519",
        "key": "31c4647554640c464c854337570217269a1fc0f8bc30591c349a410269090920"
      }
    }
  ]
}
```

Add optional filters

- `/api/v1/tokens?type=fungible` - All fungible tokens (other values are `nft` and `both` (default))
- `/api/v1/tokens?serial.number=1001` - All tokens that contain a serial number `1001` (implied that `type` will
  be `nft`)

#### Get Token by id

- Update `/api/v1/tokens/{id}` response to show NFTs by adding the `fungible` field and the `serial_numbers` list.

```json

>     {

Reviewer: nit: indentation

Author: Fixed.

  "admin_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "auto_renew_account": "0.0.6",
  "auto_renew_period": null,
  "decimals": "1000",
  "expiry_timestamp": null,
  "freeze_default": false,
  "freeze_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "fungible": false,
  "initial_supply": "2",
  "kyc_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  },
  "max_supply": "10",
  "name": "FOO COIN TOKEN",

>   "nfts": [

Author: So remove this (and in `/tokens`  as well), keep `tokens/{id}/nfts` as is?  Should I add a link to the `tokens/{id}/nfts` for applicable tokens, or is that just creating clutter?

>   "nfts": [

Reviewer 2: Yep, keep `tokens/{id}/nfts`. No, we shouldn't add links as we only have `links.next` and haven't thought through how other links would look like. Would've been better to support HATEOAS from the beginning.

>   "nfts": [

Author: Alright, that should be done now then.  Makes sense.

  },
  "symbol": "FOOCOIN",
  "token_id": "0.15.3",
  "total_supply": "2",
  "treasury_account": "0.15.10",
  "wipe_key": {
    "_type": "ProtobufEncoded",
    "key": "9c2233222c2233222c2233227d"
  }
}
```

### Token Supply distribution

- Update `/api/v1/tokens/{id}/balances` response to show `serial_numbers` when the token is an NftType token.

```json
    {
  "timestamp": "0.000002345",
  "balances": [
    {
      "account": "0.15.10",
      "nfts": [
        {
          "serial_number": 1002,
          "memo": "This is a test NFT"
        },
        {
          "serial_number": 1003,
          "memo": "This is another test NFT"
        }
      ]
    },
    {
      "account": "0.15.9",
      "nfts": [
        {
          "serial_number": 1004,
          "memo": "This is a third test NFT"
        },
        {
          "serial_number": 1005,
          "memo": "This is a fourth test NFT"
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

#### List NFTs

- GET `/api/v1/tokens/{id}/nfts` will list basic information of all NFTs for a given token.

```json
{
  "nfts": [
    {
      "created_timestamp": "1610682445.003266000",
      "memo": "This is a test NFT",

>       "serial_number": 124

Reviewer 2: Should add `modified_timestamp` and maybe `deleted` here and nfts/{id}

    }
  ],
  "links": {
    "next": null
  }
}
```

Optional Filters

> - `/api/v1/nfts?serial.number=gt:0.0.1001` - All serial numbers in range

Author: Got it, fixed.

- `/api/v1/nfts?order=desc` - All NFTs in descending order of `serial_number`
- `/api/v1/nfts?limit=x` - All NFTs taking the first `x` number of NFTs

#### Get NFT by id

- GET `/api/v1/tokens/{id}/nfts/{serialNumber}` will show information about an individual NFT.

```json
{
  "created_timestamp": "1610682445.003266000",
  "memo": "This is a test NFT",
  "serial_number": 124
}
```

#### Get NFT transaction history

> - GET `/api/v1/tokens/{id}/nfts/{serialNumber}/transactions` will return minimal transaction information for each

Reviewer 2: I'm not sure the need/value of this API. Would prefer we not add this quite different API until there's a custom ask.

Reviewer 2: I see. Does a "history of the nft" mean its full transaction history or just a history of transfers? It would be more efficient for the latter to just show nft 1 was transferred from bob to carol to Dave. Then it could be `../transfers`.

Reviewer 3: Yeah I was gonna say in terms of history I would imagine it would refer to ownership transfer in which case it should be under the `/api/v1/transfers` API I suggested.

I don't think we should load history under the `/api/v1/tokens/{id}/nfts/{serialNumber}` we already have problems with cases where API responses have information that doesn't really need to be there.
If they want history it should be an explicit API resource

Author: Per discussion with Atul just now
`The history endpoint should have a cradle-to-grave history of the NFT ( == <token_type + serial_number>)`
So it would involve minting and burning/deleting/etc.

Reviewer 3: In that case it maps across transaction types (CRYPTOTRANSFER, TOKENCREATION, TOKENFREEZE, TOKENUNFREEZE, TOKENGRANTKYC, TOKENREVOKEKYC, TOKENDELETION, TOKENUPDATE, TOKENMINT, TOKENBURN, TOKENWIPE, TOKENASSOCIATE, TOKENDISSOCIATE) which justifies `/api/v1/tokens/{id}/nfts/{serialNumber}/transactions`.

Also even though it's at the `/api/v1/tokens/{id}/nfts/{serialNumber}/transactions` endpoint we should design it in a way that supports a potential `/api/v1/tokens/{id}/transactions` for the FUNGIBLE case. Out of scope for this design but worth considering during the implementation phase.

  transaction involving a given NFT.

```json
{
  "transactions": [
    {
      "consensus_timestamp": "1618591023.997420022",
      "transaction_id": "0.0.19789-1618805680-742097949",

>       "transaction_type": "TOKENBURN"

Reviewer: should we not add the `transaction` prefix?

Author: You mean just have `id` and `type`?  I'd be okay with that.

Reviewer: yes. we had a discussion sometime back in our scrum about the redundant prefix. I think the conclusion is for new fields we should not have such redundant prefix.

Author: My only concern is that our regular `/transactions` endpoint has the prefixes.  But I'm okay with making that change here.

Author: Fixed.
